Subscription emails for customer should not link to order page if customer has no user account

kristinalim opened the pull request:
#### What? Why?

Closes #2534

Updates subscription-related emails for customers, removing link to order page for customers who do not have a user account. Adds custom text for placement emails for this scenario.

#### What should we test?

For below, make sure that the OFN instance is already [configured for Stripe integration](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Setting-up-Stripe-on-an-OFN-instance).

The following should also be set up for the shop:

1. Subscriptions are enabled (Shop Preferences > Subscriptions)
2. Stripe payment method (Shop Preferences > Payment Methods)

There are three types of subscription emails affected by this PR:

1. Placement of order
2. Confirmation of order
3. Failure to automatically charge payment

And three scenarios that should be tested:

1. Customer has no associated user account
2. Customer has user account, and shop is configured to allow customer to edit placed orders
3. Customer has user account, but shop is configured to restrict customer from editing placed orders

**Scenario 1: Customer has no associated user account**

1. Create a new ongoing order cycle (OC).
2. Create the customer. Make sure there is no user account for this email yet.
3. Create a subscription that starts before current time. Select Cheque payment method.
4. Wait approximately 5 minutes. Find the order placement email, and check that it does not contain a link to the order. _"This order was automatically created for you. This order was automatically created for you. You can make changes until orders close on Fri Aug 31 @ 11:00PM by contacting Amazing Shop via shop@example.com."_
5. To simulate the OC closing, :warning: set its close time to a couple of minutes before current time.
6. Wait approximately 5 minutes. Find the order confirmation email, and check that it does not contain a link to the order. _"This order was automatically placed for you, and it has now been finalised. If you have any questions you can contact Amazing Shop via shop@example.com."_

(AFAIK, there is no way for a customer with no user account to receive the SO failed payment for the subscription. I will ask and will update this when I get info.)

**Scenario 2: Customer has user account, and shop is configured to allow customer to edit placed orders**

(Change enterprise setting: Shop Preferences > Change Orders > "Customers can change / cancel orders while OC is open")

1. :warning: Open the OC again by updating its close date.
2. Create the customer.
3. **As customer:** Sign up with the customer email.
4. **As customer:** Save [credit card details that would charge successfully](https://stripe.com/docs/testing#cards), and give the shop permission to charge you.
5. Create a subscription that starts before current time. Select Stripe payment method.
6. Wait approximately 5 minutes. Find the order placement email, and check that it contains a link to "make changes" to the order. _"This order was automatically created for you. You can make changes until orders close on Fri Aug 31 @ 11:00PM. If you have any questions you can contact Amazing Shop via shop@example.com."_
7. To simulate the OC closing, :warning: set its close time to a couple of minutes before current time.
8. Wait approximately 5 minutes. Find the order confirmation email, and check that it contains a link to "view details" of the order. _"This order was automatically placed for you, and it has now been finalised. You can view details of this order at any time. If you have any questions you can contact Amazing Shop via shop@example.com."_
9. :warning: Open the OC again by updating its close date.
10. **As customer:** Change the saved credit card details to a one that would fail when attempting to charge.
11. Create a subscription that starts before current time. Select Stripe payment method.
12. Wait approximately 5 minutes. Find the order placement email, and check that it contains a link to "view details" of the order. _"The payment for your subscription with Amazing Shop failed because of a problem with your credit card. Amazing Shop has been notified of this failed payment. You can view details of this order at any time. If you have any questions you can contact Amazing Shop via shop@example.com."_

**Scenario 3: Customer has user account, but shop is configured to restrict customer from editing placed orders**

(Change enterprise setting: Shop Preferences > Change Orders > "Placed orders cannot be changed / cancelled")

1. :warning: Open the OC again by updating its close date.
2. Create a subscription that starts before current time. Select Stripe payment method.
3. Wait approximately 5 minutes. Find the Order placement email, and check that it contains a link to "view details" of the order. _"This order was automatically created for you. You can view details of this order at any time. If you have any questions you can contact Amazing Shop via shop@example.com."_

#### Release notes

- Remove link to Order page from Subscription emails for Customers that do not have associated User.

Changelog Category: Fixed

sauloperez commented on app/helpers/shop_mail_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module ShopMailHelper
+  def mail_long_datetime_format
```
an example of a date that follows the format below as method doc will come in handy

kristinalim replied:
Thanks, @sauloperez. Done here: fc033822984cbd3ab183fe69874540f0f47ee9c9

kristinalim replied:
Eep! :grin: After reading again, it's quite obvious that's what @sauloperez meant, and it makes a lot more sense. Sorry for the trouble, and thanks for checking! I pushed the change: ad8b120e4869c53a0ec04abff8bdd63de4191d7c PR checks remain green.

luisramos0 commented on spec/mailers/subscription_mailer_spec.rb:
```diff
@@ -39,7 +40,36 @@
+        SubscriptionMailer.placement_email(order, {}).deliver
+      end
+
+      let(:customer) { create(:customer, enterprise: shop) }
```
is there a reason to have this let after the before and not together with the others lets above?

luisramos0 commented:
@kristinalim I think you can decide what to do with the two tiny details raised above and move to test ready afterwards.

kristinalim commented:
I made the change, and also squashed a local change I hadn't pushed (oops! :worried:) which corrects `%distributor}` in the language file to `%{distributor}`. I also rebased this upon master.

Checks all good. Moving this to Test Ready now.

kristinalim commented:
@luisramos0 Eep. :grin: I updated the method doc comment already. https://github.com/openfoodfoundation/openfoodnetwork/pull/2624#discussion_r214517145

mkllnk commented:
Staged on https://staging1.openfood.com.au/.

sstead commented:
**Testing notes**

All three scenarios were behaving correctly.

https://docs.google.com/document/d/13HFUszOoEyvNqdBlefMZ5K0haXj28iMiKJaLPkxXPF4/edit#
